### ISSUE TITLE
Alt-Svc and ORIGIN

### ISSUE THREAD

**Issue by martinthomson** [httpwg/http-extensions#1691]
We should find space to discuss the implications of ORIGIN as it relates to Alt-Svc usage.  Basic idea: an alternative service fails for an origin if ORIGIN frames are received that do not list that origin.

**Comment by RyanTheOptimist** [comment]
I'm not quite sure I understand the implication here. I thought that by definition an alternative was authoritative for the origin (assuming the connection is correctly authenticated). Is that idea that a server could send an ORIGIN frame in such a way that it exclude the original origin? I thought 421 was for this use case, and ORIGIN was for addressing connection coalescing? But maybe I'm misunderstanding the proposal?

**Comment by martinthomson** [comment]
So my understanding of the positioning of ORIGIN and 421 is that 421 is reactive (you only send it in response to a request), whereas ORIGIN is proactive (you send it to pre-empt potentially unwanted requests).  Overall, the two are intended to be complementary.

**Comment by RyanTheOptimist** [comment]
I think I must be misunderstanding RFC 8336. I thought it was specifically to add additional origins in additional to the initial origin. Specifically this text:

>    When an ORIGIN frame is first received and successfully processed by
>    a client, the connection's Origin Set is defined to contain an
>    initial origin.  
>    ...
>    The contents of that ORIGIN frame (and subsequent ones) allow the
>    server to incrementally add new origins to the Origin Set, as
>    described in Section 2.2.

As I read that, there is no mechanism for the ORIGIN frame to *remove* origins, only to add them. How does the ORIGIN frame cause origins to be removed from the Origin Set?

**Comment by martinthomson** [comment]
The introduction says it fairly clearly: clients might coalesce origins onto a connection based on the certificate.  A certificate, particularly one with a bunch of wildcards (*.example.com) can be rather permissive.  ORIGIN allows a server to say which specific domains are active.

For instance, you connect to www.example.com and it has a certificate for *.example.com.  You might at that point coalesce assets.example.com onto that connection, not knowing if the server supports ORIGIN.  If the server sent ORIGIN(www.example.com) then you would not coalesce any other names unless the server subsequently sends more ORIGIN frames, at which point you can send requests for any origin thus added.

The removal of origins therefore occurs implicitly along with the first ORIGIN frame.

**Comment by RyanTheOptimist** [comment]
Since this issue was discussion alt-svc, I thought the concern was the ORIGIN frame making the connection no longer valid for the origin that it was initially established for (which the ORIGIN frame is unable to do). But I think you're saying the the concern is what happens when a connection to an alternative service (origin A) is coalesced onto an existing connection (to origin B) which had a cert valid for multiple origins (A & B) but the ORIGIN frame removes origin A. Is that the concern this issue describes?

**Comment by martinthomson** [comment]
Yes.  Say you get alt-svc for example.com that refers to cdn.example.  You already have a connection active to cdn.example (maybe with some other target name) and the IP address matches (if you care about that).  Receiving an ORIGIN frame that did not include example.com would mean that THAT alternative cannot be used and a fresh connection might be needed.

**Comment by martinthomson** [comment]
Let's say that the cdn.example connection was originally established to a different host.  Then example.com advertises uses the IP address of cdn.example in Alt-Svc.  This IP address does not appear in A/AAAA records for example.com, so the client would never have attempted to use that connection, even though the offered certificate lists example.com.  However, once Alt-Svc points to that IP address, it can then use that connection for example.com.

The Alt-Svc spec isn't very clear about whether coalescing can be used.  This is also a part of fetch that isn't well mapped out.  I model this process as "obtain a connection to the origin" rather than "establish a connection to the origin".  In that framing, using an existing connection is a perfectly valid choice.

If you only ever use Alt-Svc for the purposes of creating connections (and not reusing them), then ORIGIN can't disqualify a conneciton.

**Comment by martinthomson** [comment]
I've opened #1696 to track what is probably a bigger issue.  We might want to fix these two issues with the same text.

**Comment by MikeBishop** [comment]
@RyanTheOptimist, there are two directions ORIGIN can go.  The one originally contemplated in the spec is that ORIGIN will specify a tighter set of origins than the cert presented on the connection would suggest; the server is able to prevent requests that the client might otherwise make.

In the now-dead Secondary Certs draft, a second use for ORIGIN was proposed -- the server could also indicate additional origins not on the primary cert, and the client could ask to see the appropriate certs if interested.  However, with that draft gone, ORIGIN is back to only restricting, not expanding, the set of possible origins.  If the server indicates origins not on the certificate, the client will consider the connection not authoritative for those names and not make requests (or accept pushes) for them.
